pydantic_ai_slim/pydantic_ai/ui/vercel_ai/_adapter.py

```diff
@@ -197,3 +215,226 @@ def load_messages(cls, messages: Sequence[UIMessage]) -> list[ModelMessage]:  #
+                        result.append(UIMessage(id=_message_id_generator(), role='user', parts=user_ui_parts))
+
+            elif isinstance(  # pragma: no branch - All message types are covered (no tests for empty ModelResponse)
+                msg, ModelResponse
+            ):
+                ui_parts: list[UIMessagePart] = []
+                text_parts: list[str] = []
+                had_interruption = False
+
+                # For builtin tools, returns can be in the same ModelResponse as calls
+                # Build a local mapping for this message
+                local_builtin_returns: dict[str, BuiltinToolReturnPart] = {}
+                for part in msg.parts:
+                    if isinstance(part, BuiltinToolReturnPart):
+                        local_builtin_returns[part.tool_call_id] = part
+
+                for part in msg.parts:
+                    if isinstance(part, BuiltinToolReturnPart):
+                        # Skip builtin tool returns - they're handled by the tool call logic
+                        continue
+                    elif isinstance(part, TextPart):
+                        # If this is the first text after an interruption, prepend separator
+                        if had_interruption:
+                            text_parts.append('\n\n' + part.content)
+                        else:
+                            text_parts.append(part.content)
+                    elif isinstance(part, ThinkingPart):
+                        if text_parts:
+                            ui_parts.append(TextUIPart(text=''.join(text_parts), state='done'))
+                            text_parts = []
+                            had_interruption = False
+                        ui_parts.append(ReasoningUIPart(text=part.content, state='done'))
+                    elif isinstance(part, FilePart):
+                        if text_parts:
+                            ui_parts.append(TextUIPart(text=''.join(text_parts), state='done'))
+                            text_parts = []
+                            had_interruption = False
+                        ui_parts.append(
+                            FileUIPart(
+                                url=part.content.data_uri,
+                                media_type=part.content.media_type,
+                            )
+                        )
+                    elif isinstance(part, BaseToolCallPart):  # pragma: no branch - All assistant part types are covered
+                        if text_parts:
+                            ui_parts.append(TextUIPart(text=''.join(text_parts), state='done'))
+                            text_parts = []
+
+                        # Mark that we had an interruption for next text part
+                        had_interruption = True
+
+                        if isinstance(part, BuiltinToolCallPart):
+                            prefixed_id = _make_builtin_tool_call_id(part.provider_name, part.tool_call_id)
```

Comment:
What do we need this for?

Reply:
I honestly just read it in [your review](https://github.com/pydantic/pydantic-ai/pull/3068/files#r2402440610) of 3068 and thought I'd need a helper as well, but I'm only calling it once